### ISSUE TITLE
Panic with "poll_response called after response returned"

### ISSUE THREAD

**Issue by nox** [hyperium/h2#524]
We have a panic deep in h2, I looked into it for a bit and AFAIK the panic message is wrong and the future did not get polled twice and this is instead a protocol bug where h2 popped an event which it expected to be a headers event but it got data or trailers instead.

https://github.com/hyperium/h2/blob/bcaaaf6dd98d8da69e4755b0bcb3e0f4dd3cfee6/src/proto/streams/recv.rs#L302-L313

I concluded that the panic didn't actually happen because the future was polled twice because the future is only ever polled by Hyper from behind a `futures_util::future::future::Map<T>`, as shown in the attached stacktrace (see below):

```
4   h2              h2::client::ResponseFuture::poll (client.rs:1300)
5   futures_util    [inlined] futures_util::future::future::map::Map<T>::poll (map.rs:55)
6   futures_util    futures_util::future::future::Map<T>::poll (lib.rs:102)
```

That mapping is done in Hyper itself:

https://github.com/hyperium/hyper/blob/237b2ce08341266b62a8d1cfa974779c511a0710/src/proto/h2/client.rs#L269-L288

In this snippet, `fut` in `fut.map` is the `ResponseFuture` whose `poll` method panicked, as you can see there is no opportunity to poll it twice, as `<Map<T>>::poll` would have panicked already:

https://github.com/rust-lang/futures-rs/blob/2fa7babdf92e4e6b1e874054a2773c5c330a3299/futures-util/src/future/future/map.rs#L53-L64

<hr><details><summary><em>Click to expand stacktrace</em></summary>

```
Application Specific Information:
poll_response called after response returned

Thread 0 Crashed:
0   std             std::sys_common::backtrace::__rust_end_short_backtrace (backtrace.rs:141)
1   std             std::panicking::begin_panic (panicking.rs:519)
2   h2              h2::proto::streams::recv::Recv::poll_response (recv.rs:306)
3   h2              h2::proto::streams::streams::OpaqueStreamRef::poll_response (streams.rs:1237)
4   h2              h2::client::ResponseFuture::poll (client.rs:1300)
5   futures_util    [inlined] futures_util::future::future::map::Map<T>::poll (map.rs:55)
6   futures_util    futures_util::future::future::Map<T>::poll (lib.rs:102)
7   hyper           [inlined] hyper::client::dispatch::Callback<T>::send_when::{{closure}} (dispatch.rs:243)
8   futures_util    futures_util::future::poll_fn::PollFn<T>::poll (poll_fn.rs:55)
9   tokio           [inlined] tokio::runtime::task::core::CoreStage<T>::poll::{{closure}} (core.rs:235)
10  tokio           [inlined] tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut (unsafe_cell.rs:14)
11  tokio           [inlined] tokio::runtime::task::core::CoreStage<T>::poll (core.rs:225)
12  tokio           [inlined] tokio::runtime::task::harness::poll_future::{{closure}} (harness.rs:422)
13  std             std::panic::AssertUnwindSafe<T>::call_once (panic.rs:322)
14  std             [inlined] std::panicking::try::do_call (panicking.rs:379)
15  std             [inlined] std::panicking::try (panicking.rs:343)
16  std             [inlined] std::panic::catch_unwind (panic.rs:396)
17  tokio           tokio::runtime::task::harness::poll_future (harness.rs:409)
18  tokio           [inlined] tokio::runtime::task::harness::Harness<T>::poll_inner (harness.rs:89)
19  tokio           tokio::runtime::task::harness::Harness<T>::poll (harness.rs:59)
20  tokio           [inlined] tokio::runtime::task::raw::RawTask::poll (raw.rs:66)
21  tokio           [inlined] tokio::runtime::task::Notified<T>::run (mod.rs:171)
22  tokio           [inlined] tokio::runtime::thread_pool::worker::Context::run_task::{{closure}} (worker.rs:370)
23  tokio           [inlined] tokio::coop::with_budget::{{closure}} (coop.rs:121)
24  std             [inlined] std::thread::local::LocalKey<T>::try_with (local.rs:272)
25  std             std::thread::local::LocalKey<T>::with (local.rs:248)
26  tokio           [inlined] tokio::coop::with_budget (coop.rs:114)
27  tokio           [inlined] tokio::coop::budget (coop.rs:98)
28  tokio           tokio::runtime::thread_pool::worker::Context::run_task (worker.rs:348)
29  tokio           tokio::runtime::thread_pool::worker::Context::run (worker.rs:318)
30  tokio           [inlined] tokio::runtime::thread_pool::worker::run::{{closure}} (worker.rs:303)
31  tokio           tokio::macros::scoped_tls::ScopedKey<T>::set (scoped_tls.rs:61)
32  tokio           tokio::runtime::thread_pool::worker::run (worker.rs:300)
33  tokio           [inlined] tokio::runtime::thread_pool::worker::Launch::launch::{{closure}} (worker.rs:279)
34  tokio           [inlined] tokio::runtime::blocking::task::BlockingTask<T>::poll (task.rs:42)
35  tokio           [inlined] tokio::runtime::task::core::CoreStage<T>::poll::{{closure}} (core.rs:235)
36  tokio           tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut (unsafe_cell.rs:14)
37  tokio           [inlined] tokio::runtime::task::core::CoreStage<T>::poll (core.rs:225)
38  tokio           [inlined] tokio::runtime::task::harness::poll_future::{{closure}} (harness.rs:422)
39  std             [inlined] std::panic::AssertUnwindSafe<T>::call_once (panic.rs:322)
40  std             [inlined] std::panicking::try::do_call (panicking.rs:379)
41  std             std::panicking::try (panicking.rs:343)
42  std             [inlined] std::panic::catch_unwind (panic.rs:396)
43  tokio           [inlined] tokio::runtime::task::harness::poll_future (harness.rs:409)
44  tokio           [inlined] tokio::runtime::task::harness::Harness<T>::poll_inner (harness.rs:89)
45  tokio           tokio::runtime::task::harness::Harness<T>::poll (harness.rs:59)
46  tokio           [inlined] tokio::runtime::task::raw::RawTask::poll (raw.rs:66)
47  tokio           [inlined] tokio::runtime::task::Notified<T>::run (mod.rs:171)
48  tokio           [inlined] tokio::runtime::blocking::pool::Inner::run (pool.rs:278)
49  tokio           [inlined] tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}} (pool.rs:258)
50  std             std::sys_common::backtrace::__rust_begin_short_backtrace (backtrace.rs:125)
51  std             [inlined] std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}} (mod.rs:474)
52  std             [inlined] std::panic::AssertUnwindSafe<T>::call_once (panic.rs:322)
53  std             [inlined] std::panicking::try::do_call (panicking.rs:379)
54  std             [inlined] std::panicking::try (panicking.rs:343)
55  std             [inlined] std::panic::catch_unwind (panic.rs:396)
56  std             [inlined] std::thread::Builder::spawn_unchecked::{{closure}} (mod.rs:473)
57  core            core::ops::function::FnOnce::call_once{{vtable.shim}} (function.rs:227)
58  alloc           [inlined] alloc::boxed::Box<T>::call_once (boxed.rs:1328)
59  alloc           [inlined] alloc::boxed::Box<T>::call_once (boxed.rs:1328)
60  std             std::sys::unix::thread::Thread::new::thread_start (thread.rs:71)
```
</details>

**Comment by nox** [comment]
It is a bit unfortunate that there is no way to see if the event that caused the panic was an `Event::Headers(Server(_))` or something completely different.

**Comment by nox** [comment]
In #521 h2 started ignoring informational frames.

https://github.com/hyperium/h2/blob/bcaaaf6dd98d8da69e4755b0bcb3e0f4dd3cfee6/src/proto/streams/recv.rs#L229-L239

AFAIK this code can be reached from a stream whose state was `Idle`, which means that we can move a stream's state from `Idle` to `Open { local: AwaitingHeaders, remote: Streaming }` without pushing the `Event::Headers` that caused the state change.

https://github.com/hyperium/h2/blob/bcaaaf6dd98d8da69e4755b0bcb3e0f4dd3cfee6/src/proto/streams/recv.rs#L164

https://github.com/hyperium/h2/blob/89d91b0a4f879bede3d3348bed061d33544f2682/src/proto/streams/state.rs#L144-L155

I'm wondering if the stream can be then passed to `recv_data` with this `Open { remote: Streaming, .. }` and push an `Event::Data` even though we still did not push `Event::Headers`.

This wouldn't trigger:

https://github.com/hyperium/h2/blob/bcaaaf6dd98d8da69e4755b0bcb3e0f4dd3cfee6/src/proto/streams/recv.rs#L542-L552

Because `is_recv_streaming` returns true:

https://github.com/hyperium/h2/blob/89d91b0a4f879bede3d3348bed061d33544f2682/src/proto/streams/state.rs#L385-L393

So we end up pushing an `Event::Data`, which AFAIU would be the first event pushed ever for that stream, making `ResponseFuture::poll` panic:

https://github.com/hyperium/h2/blob/bcaaaf6dd98d8da69e4755b0bcb3e0f4dd3cfee6/src/proto/streams/recv.rs#L627-L629

**Comment by nox** [comment]
Fixed in #527.
